src/xwayland/xserver.rs

```diff
@@ -1,509 +1,382 @@
-    Interest, LoopHandle, Mode,
-};
-use wayland_server::{
-    backend::{ClientData, ClientId, DisconnectReason},
-    Client, DisplayHandle,
-};
+use tracing::{info, trace};
+use wayland_server::backend::ClientData;
+use wayland_server::backend::DisconnectReason;
+use wayland_server::{Client, DisplayHandle};
 
-use tracing::{error, info, instrument};
+use crate::{utils::user_data::UserDataMap, wayland::compositor::CompositorClientState};
 
 use super::x11_sockets::{prepare_x11_sockets, X11Lock};
-use crate::utils::user_data::UserDataMap;
-#[cfg(feature = "wayland_frontend")]
-use crate::wayland::compositor::CompositorClientState;
 
-/// The XWayland handle
+/// A handle to a running XWayland process. Using XWayland as an xserver for
+/// X11-based clients requires two connections: one wayland socket, where
+/// XWayland creates surfaces for its clients, and one X11 socket, where a
+/// compositor's implementation of an X11 window manager connects and handles
+/// X11 events.
+///            ┌───────────────────┐
+///            │                   │
+///            │                   │ X11 Clients
+///            │     Xwayland      │ ◄─────
+///            │                   │
+///            │                   │
+///            └──┬──────────────▲┘
+///               │              │
+/// Wayland Socket│              │ X11 Socket
+///               │              │
+///               │              │
+///  ┌────────────▼─────────────┼────────────┐
+///  │  Compositor               │            │
+///  │                         ┌─┴───────┐    │
+///  │                         │ XWM     │    │
+///  │                         │         │    │
+///  │                         └─────────┘    │
+///  │                                        │
+///  └────────────────────────────────────────┘
+///
+/// This struct handles integrating the XWayland process itself into the event
+/// loop, but a [separate X11 window manager implementation](crate::xwayland::xwm::X11Wm)
+/// is needed as well to support X11 clients.
+///
+/// To shut down the instance, dropping this handle is generally sufficient,
+/// along with the X11 connection that was passed to the window manager
+/// implementation, if any. The process will die once the connections to it
+/// are closed.
 #[derive(Debug)]
 pub struct XWayland {
-    inner: Arc<Mutex<Inner>>,
+    inner: Arc<Mutex<Instance>>,
+    source: calloop::generic::Generic<calloop::generic::FdWrapper<RawFd>>,
+
+    // So we can disconnect the client on drop.
+    dh: DisplayHandle,
+    client: Client,
 }
 
-/// Events generated by the XWayland manager
-///
-/// This is a very low-level interface, only notifying you when the connection
-/// with XWayland is up, or when it terminates.
-///
-/// Your WM code must be able to handle the XWayland server connecting then
-/// disconnecting several time in a row, but only a single connection will
-/// be active at any given time.
+/// Events generated by an [XWayland] instance.
```

Comment:
```suggestion
/// Events generated by an [`XWayland`] instance.
```

Reply:
The backticks aren't necessary, and it looks worse to me, but I can change it if you really want.